### PR TITLE
[improve][build] PIP-156 Build Pulsar Server on Java 17

### PR DESCRIPTION
**Pull request by heesung-sn**
Master Issue: #15207

### Motivation

This is to upgrade Pulsar Server's Java build version to 17.
The detail motivation is described in the PIP-156, https://github.com/apache/pulsar/issues/15207

There will be multiple commits under this PR to complete the tasks under PIP-156.

### Modifications
This changes Pulsar Server modules' javac release version to 17 except client (shared) modules
The detail modification is described in the PIP-156, https://github.com/apache/pulsar/issues/15207

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change added tests and can be verified as follows:
  - *will add integration tests for different java runtime versions if currently not covered*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API: (yes / **no**)
  - The schema: (yes / **no** / don't know)
  - The default values of configurations: (yes / **no**)
  - The wire protocol: (yes / **no**)
  - The rest endpoints: (yes / **no**)
  - The admin cli options: (yes / **no**)
  - Anything that affects deployment: (**yes** / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [x] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [ ] `no-need-doc` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

**Review by nicoloboschi — CHANGES_REQUESTED** [review]
We must create a matrix test suite where we run the Pulsar client suite with jdk8 and jdk17 to ensure compatibility

File: tests/integration/src/test/java/org/apache/pulsar/tests/integration/io/sinks/KinesisSinkTester.java
Comment (heesung-sn) [reply tests/integration/src/test/java/org/apache/pulsar/tests/integration/io/sinks/KinesisSinkTester.java:197]: I think this is related to this pull. 
It seems like the java17 compiled KinesisSink code saves more cpu cycles than before and changed the initial KPL behavior - now we can see the KPL default record aggregation in this test, which was not covered in the original test.

File: build/run_unit_group.sh
Comment (heesung-sn) [reply build/run_unit_group.sh:81]: We have only `ManagedLedgerCompressionTest` which tags `broker-jdk17`(and `broker` too).  I don't mind removing this `broker_jdk17` test as it is already covered in the broker test, but it might be useful in the future when we start supporting jdk18+, as a broker jdk base test, just like the original broker jdk8 test. 

```
@Test(groups = {"broker", "broker-jdk17"})
public class ManagedLedgerCompressionTest extends BrokerTestBase {
```

**Comment by heesung-sn** [comment]
> We must create a matrix test suite where we run the Pulsar client suite with jdk8 and jdk17 to ensure compatibility

Yes, that is a part of the plan. I was focusing on the build upgrade and regression checks first. After I confirm that the existing tests pass, I will start to add the `Pulsar client jdk8 and jdk17 compatibility test` commits here.

File: pulsar-client-cpp/docker-tests.sh
Comment (heesung-sn) [reply pulsar-client-cpp/docker-tests.sh]: Added apachepulsar/pulsar-build:ubuntu-16.04-pb3-jdk17 with jdk17 pre-installed. I updated the script to use it instead.

**Review by merlimat — APPROVED** [review]

**Comment by lhotari** [comment]
@heesung-sn Just that you know about it, it's usually useful to work on GitHub Actions workflow changes in your own fork before opening a PR to apache/pulsar. Please check the comments in https://github.com/apache/pulsar/pull/15385#issuecomment-1113050807 .
When running builds in your own fork, you can also get ssh access to the build VMs to investigate issues. ssh access is disabled because of security reasons in apache/pulsar builds.

File: README.md
Comment (lhotari) [inline README.md:100]: For Broker, the recommended runtime environment has been Java 11 since Pulsar 2.8.0 release.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ManagedLedgerCompressionTest.java
Comment (lhotari) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ManagedLedgerCompressionTest.java:37]: just remove it. The reason why this specific test was marked for broker-jdk8 was that there's a change between Java 8 and Java 11 that showed up in this test. It's now obsolete, so remove the broker-jdk8 group completely.

File: tests/integration/src/test/java/org/apache/pulsar/tests/integration/io/sinks/KinesisSinkTester.java
Comment (lhotari) [inline tests/integration/src/test/java/org/apache/pulsar/tests/integration/io/sinks/KinesisSinkTester.java:245]: Please remove this from this PR. Submit as a separate PR.

File: pom.xml
Comment (lhotari) [inline pom.xml:82]: I think it would be more consistent to use `compiler` instead of `javac`, thus renaming `pulsar.client.javac.release` -> `pulsar.client.compiler.release`.

**Comment by lhotari** [comment]
> There will be multiple commits under this PR to complete the tasks under PIP-156.

I don't think that it's a good goal to complete all changes for PIP-156 in a single PR. Smaller PRs are easier to review and get merged. For example, there's no sense in putting the Kinesis related changes in this PR. That should be solved separately.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ManagedLedgerCompressionTest.java
Comment (heesung-sn) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ManagedLedgerCompressionTest.java]: removed.

File: pom.xml
Comment (heesung-sn) [reply pom.xml]: updated.

**Comment by heesung-sn** [comment]
Hi @lhotari @nicoloboschi,
I think I resolved your comments in the latest push. Could you review this PR again by any chance?

File: README.md
Comment (nicoloboschi) [inline README.md:88]: 2.11 is not released. I believe it is better to keep the this table only up to 2.10.0, then update it when 2.11 will be publicly released

File: README.md
Comment (nicoloboschi) [inline README.md:164]: what about branch-2.9 and 2.10 ?

File: README.md
Comment (nicoloboschi) [inline README.md:86]: I believe we have to write much clear that the Pulsar IO and Functions will run on jdk17 at runtime. Here and in the website

**Review by nicoloboschi — CHANGES_REQUESTED** [review]
Good work!

File: README.md
Comment (heesung-sn) [reply README.md]: updated.

File: README.md
Comment (heesung-sn) [reply README.md]: I made this part to the master branch only. (in order to clarify, after this commit, the master branch needs to run on java17).
I agree that we need to revisit this table again for 2.11 release too.

File: README.md
Comment (heesung-sn) [reply README.md:86]: updated.

**Review by nicoloboschi — APPROVED** [review]

**Comment by merlimat** [comment]
@lhotari PTAL again.

**Review by momo-jun — COMMENTED** [review]
There are three more occurrences of the JRE/JDK versions (8 or 11) in the following topics. Should they be updated as well?
- deploy-bare-metal.md 
image

- deploy-bare-metal-multi-cluster.md
- getting-started-standalone.md

**Comment by heesung-sn** [comment]
> There are three more occurrences of the JRE/JDK versions (8 or 11) in the following topics. Should they be updated as well?
> 
> - deploy-bare-metal.md 
> 
> image
> 
> 
> 
> - deploy-bare-metal-multi-cluster.md
> 
> - getting-started-standalone.md

Thank you for this catch. Yes, I think we need to update these parts too. If you dont mind, I will probably take this as a separate PR for further doc updates. I think we probably start scoping this PR for code change only to avoid the repeated CI test runs.

**Review by lhotari — CHANGES_REQUESTED** [review]
@heesung-sn I reviewed the changes and since it would have been too hard to explain the details, I have push changes to an experiment that I'm running in my fork, it's https://github.com/lhotari/pulsar/pull/69 . 
I simplified the pulsar-ci.yaml changes to just this change:
https://github.com/lhotari/pulsar/pull/69/commits/5521c543737b5aa0542f30d80c9bb7f3a19b2b2d

Instead of adding a new job, it was sufficient to extend the existing integration-tests matrix job and add a matrix property "runtime_jdk" 
example
```
          - name: Shade on Java 8
            group: SHADE_RUN
            runtime_jdk: 8
            setup: ./build/run_integration_group.sh SHADE_BUILD

          - name: Shade on Java 11
            group: SHADE_RUN
            runtime_jdk: 11
            setup: ./build/run_integration_group.sh SHADE_BUILD

          - name: Shade on Java 17
            group: SHADE_RUN
            setup: ./build/run_integration_group.sh SHADE_BUILD
```

There was already a feature to run a setup command. 

this would be the change to support the `runtime_jdk` matrix property:
```
      - name: Set up runtime JDK ${{ matrix.runtime_jdk }}
        uses: actions/setup-java@v2
        if: ${{ matrix.runtime_jdk && needs.changed_files_job.outputs.docs_only != 'true' }}
        with:
          distribution: 'temurin'
          java-version: ${{ matrix.runtime_jdk }}
```

If the build runs fine in my fork, we could also proceed by synchronizing the changes to this PR. Are you fine with that?

**Comment by heesung-sn** [comment]
> @heesung-sn I reviewed the changes and since it would have been too hard to explain the details, I have push changes to an experiment that I'm running in my fork, it's [lhotari#69](https://github.com/lhotari/pulsar/pull/69) . I simplified the pulsar-ci.yaml changes to just this change: [lhotari@5521c54](https://github.com/lhotari/pulsar/commit/5521c543737b5aa0542f30d80c9bb7f3a19b2b2d)
> 
> Instead of adding a new job, it was sufficient to extend the existing integration-tests matrix job and add a matrix property "runtime_jdk" example
> 
> ```
>           - name: Shade on Java 8
>             group: SHADE_RUN
>             runtime_jdk: 8
>             setup: ./build/run_integration_group.sh SHADE_BUILD
> 
>           - name: Shade on Java 11
>             group: SHADE_RUN
>             runtime_jdk: 11
>             setup: ./build/run_integration_group.sh SHADE_BUILD
> 
>           - name: Shade on Java 17
>             group: SHADE_RUN
>             setup: ./build/run_integration_group.sh SHADE_BUILD
> ```
> 
> There was already a feature to run a setup command.
> 
> this would be the change to support the `runtime_jdk` matrix property:
> 
> ```
>       - name: Set up runtime JDK ${{ matrix.runtime_jdk }}
>         uses: actions/setup-java@v2
>         if: ${{ matrix.runtime_jdk && needs.changed_files_job.outputs.docs_only != 'true' }}
>         with:
>           distribution: 'temurin'
>           java-version: ${{ matrix.runtime_jdk }}
> ```
> 
> If the build runs fine in my fork, we could also proceed by synchronizing the changes to this PR. Are you fine with that?

Thank you for further simplifying this CI test workflow change. I am ok with that.

**Review by lhotari — APPROVED** [review]
LGTM

**Comment by lhotari** [comment]
> Thank you for further simplifying this CI test workflow change. I am ok with that.

@heesung-sn I have synchronized my changes to this PR. 

I reverted the change that switched to use `ubuntu-16.04-pb3-jdk17` for the pulsar-client-cpp build image . That conflicts with PIP-155 #15376 changes that @merlimat implemented.
> Update build image to ubuntu:20.04 (There is no Python 3.7 support in the old Ubuntu)

**Comment by lhotari** [comment]
There's also a need to simplify the pom.xml . Previous the build has been compatible with Java 8 & Java 11 and that has been handled with the jdk11 maven profile. Since Java 17+ will be required for building, it would be useful to remove the jdk11 maven profile solution and instead use a solution that enforces Java 17+ for building (could use https://maven.apache.org/enforcer/maven-enforcer-plugin/). 
I think it's better to handle this change in a follow-up PR. I propose that we merge this current PR as-is.

**Comment by lhotari** [comment]
@heesung-sn The changes to .asf.yaml will have to be handled separately. To prevent issues where currently in-progress PRs get blocked by the new required checks (`CI - Integration - Shade on Java 8`, `CI - Integration - Shade on Java 11` and `CI - Integration - Shade on Java 17`), it's better to postpone the changes to `.asf.yaml`. The changes to `.asf.yaml` can be made after this PR has been merged for a few days.

#15496 is needed as a preparation for allowing to merge this current PR. That removes the existing required checks `CI - Unit - Brokers - JDK 8` and `CI - Integration - Shade` which are missing after the changes to the workflow. After that is merged, we are ready to merge the changes.

**Comment by lhotari** [comment]
> There's also a need to simplify the pom.xml . Previous the build has been compatible with Java 8 & Java 11 and that has been handled with the jdk11 maven profile. Since Java 17+ will be required for building, it would be useful to remove the jdk11 maven profile solution and instead use a solution that enforces Java 17+ for building (could use https://maven.apache.org/enforcer/maven-enforcer-plugin/). I think it's better to handle this change in a follow-up PR. I propose that we merge this current PR as-is.

@heesung-sn Are you interested in a follow up PR to address the above comment?

**Comment by lhotari** [comment]
> @heesung-sn Are you interested in a follow up PR to address the above comment?

I already created #15505 to address this.

**Comment by eolivelli** [comment]
@lhotari @heesung-sn

Why are we requiring JDK17 to build ?

I agree that we can ensure that Pulsar works well on JDK17 and that we should run CI on JDK17.
but in many companies it is not allowed to use JDKs that are "too modern."

can we REQUIRE JDK11 instead of JDK17 ?

**Comment by heesung-sn** [comment]
> > @heesung-sn Are you interested in a follow up PR to address the above comment?
> 
> I already created #15505 to address this.

I see this followup change. Thank you for taking care of this.

**Comment by heesung-sn** [comment]
> > Why are we requiring JDK17 to build ?
> 
> Initially, the motivation for this was to enable the use of new language features in Pulsar server-side modules, for example, using the Java record types.
> 
> > I agree that we can ensure that Pulsar works well on JDK17 and that we should run CI on JDK17. but in many companies it is not allowed to use JDKs that are "too modern."
> 
> I doubt that it's a problem for Pulsar broker. You must run Pulsar on a recent JDK version in any case, and requiring Java 17 makes sense for the broker.
> 
> Pulsar client and Pulsar admin client libraries will continue to support Java 8. For some reason, PIP-156 was specified in a way that Pulsar CLI tools also require Java 17 to run. @heesung-sn Can you explain why Java 17 is required for Pulsar CLI tools?
> 
> > can we REQUIRE JDK11 instead of JDK17 ?
> 
> That's possible, but I don't think that there's a specific reason to stop at JDK11 now that JDK17 is the current Java LTS release.

As discussed in the pip-156, https://github.com/apache/pulsar/issues/15207, the goal of this upgrade is to enable pulsar to adapt new java features. Regarding CLI java version, because CLI has dependencies on broker modules, it needs to follow broker's java version.

**Comment by lhotari** [comment]
> Regarding CLI java version, because CLI has dependencies on broker modules, it needs to follow broker's java version.

@heesung-sn Why does the CLI depend on broker modules? Would it be possible to break these dependencies so that the CLI could run on older Java versions?

**Comment by eolivelli** [comment]
In my opinion such tools that use Broker internals are very version specific and they are usually executed from the same package (and machines?) of the broker (I would never use the Compator tool of a older version or from a newer version for instance)
So it is fine to require the same jdk as for the  broker.

**Comment by michaeljmarshall** [comment]
Note that the compactor tool is accessible via the script named `bin/pulsar`. I would probably view jars activated by that script as server side jars. Note that the pulsar client jar and the pulsar admin client jar are accessible via the CLI using `bin/pulsar-client` and `bin/pulsar-admin`, respectively.

**Comment by momo-jun** [comment]
> > There are three more occurrences of the JRE/JDK versions (8 or 11) in the following topics. Should they be updated as well?
> > 
> > * deploy-bare-metal.md
> > 
> > image
> > 
> > * deploy-bare-metal-multi-cluster.md
> > * getting-started-standalone.md
> 
> Thank you for this catch. Yes, I think we need to update these parts too. If you dont mind, I will probably take this as a separate PR for further doc updates. I think we probably start scoping this PR for code change only to avoid the repeated CI test runs.

Hi Heesung, will you like to commit a follow-up PR with the remaining doc changes?

**Comment by heesung-sn** [comment]
> @heesung-sn Just that you know about it, it's usually useful to work on GitHub Actions workflow changes in your own fork before opening a PR to apache/pulsar. Please check the comments in [#15385 (comment)](https://github.com/apache/pulsar/pull/15385#issuecomment-1113050807) . When running builds in your own fork, you can also get ssh access to the build VMs to investigate issues. ssh access is disabled because of security reasons in apache/pulsar builds.

Made a PR for copying the instruction to the contribution page, https://github.com/apache/pulsar/pull/15535

**Comment by heesung-sn** [comment]
Fixed one module: https://github.com/apache/pulsar/pull/19991

**Pull request merged**
